### PR TITLE
Torch

### PR DESCRIPTION
**Pull request by stroblme**
- added and manually optimized pytorch autoencoder (tensorflow AE can be launched by triggering dedicated pipeline)
- splitted single pipeline into preprocessing, training and postprocessing
- added poetry
- added ssim metric replacing the pixel-wise accuracy
- switched from matplotlib to plotly framework

### CODE REVIEW

File: conf/base/catalog.yml [hunk 1]
```diff
@@ -5,35 +5,35 @@
...
   backend: dill
-data_preprocessing.loss_curve:
-  type: kedro.extras.datasets.matplotlib.MatplotlibWriter
-  filepath: data/08_reporting/loss_curve.png
-data_preprocessing.accuracy_curve:
-  type: kedro.extras.datasets.matplotlib.MatplotlibWriter
-  filepath: data/08_reporting/accuracy_curve.png
+loss_curve:
+  type: kedro.extras.datasets.plotly.JSONDataSet
+  filepath: data/08_reporting/loss_curve.json
```
Comment (eileen-kuehn) [inline conf/base/catalog.yml:33]: The usability of loss and accuracy curves have changed here. The intention for those data was to have an asset for reporting that can directly be used. Having a plotly json file, another step is required to be able to use it in whatever kind of report is used at this point. So we should either document how to transform this into a usable asset or move this file (to 07??) and create an additional one in the context of reporting.

File: conf/base/parameters/training.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+training:
+  number_of_features: 8
+  seed: 1337
+  epochs: 800
+  batch_size: 32 
+  ssim_filter_size: 11 
+  ssim_sigma: 1.5
+  ssim_k1: 0.01
+  ssim_k2: 0.03
```
Comment (eileen-kuehn) [inline conf/base/parameters/training.yml:9]: What are those properties about?

File: src/qc_data_preparation/pipelines/preprocessing/__init__.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+"""
+This is a boilerplate pipeline 'mnist_processing'
```
Comment (eileen-kuehn) [inline src/qc_data_preparation/pipelines/preprocessing/__init__.py:2]:
```suggestion
This is a boilerplate pipeline 'preprocessing'
```

**Review by eileen-kuehn — COMMENTED** [review]
Thanks for that great addition. Just a rough check for now, a more thorough one will follow after the static checks are passing.

File: conf/base/parameters/training.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+training:
+  number_of_features: 8
+  seed: 1337
+  epochs: 800
```
[the remark this answers is not in the document]
Comment (stroblme) [reply conf/base/parameters/training.yml:4]: with the latest autoencoder model, performance still increases after 500 epochs

[the remark this answers is not in the document]
Comment (eileen-kuehn) [reply conf/base/parameters/training.yml:4]: For both models or only one? Because otherwise we might consider varying number of epochs per model.

[the remark this answers is not in the document]
Comment (stroblme) [reply conf/base/parameters/training.yml:4]: For the tf model, the validation ssim stays more or less constant, while the training ssim increases. In case of the pt model, both the training and validation ssim increases.
Maybe we can rework the tf model to use conv.- instead of fc layers, as treating both as similar models isn't probably the best approach (by now).
However if we assimilate the structure of both models, I would expect similar optimal hyperparams. for both, although the actual results might vary a bit due to the differences between pt and tf.
If we later want to go for different architectures, then I agree; we should then rework the way we treat hyperparameters for each model, but for now, imo., this is not necessary.

File: pyproject.toml [hunk 5]
```diff
@@ -20,3 +20,45 @@ addopts = """
...
+
+[tool.black]
+line-length = 88
+include = '\.pyi?$'
+exclude = '''
+/(
+    \.git
+  | \.venv
+)/
+'''
```
Comment (eileen-kuehn) [inline pyproject.toml:64]: Please drop this :)

```suggestion
```

File: conf/base/catalog.yml (hunk 1 above)
Comment (stroblme) [reply conf/base/catalog.yml]: I added html export by introducing a separate `PlotlyDataSet` that overwrites `_save` in `JsonDataSet` and stores both, a .json file *and* .html. When reading from this dataset, the .json file is being used, as plotly currently does not provide an option to generate figures from html.
Unfortunately, Kedro-Viz does not display the figure in the dashboard which is maybe because they are expecting a strict type of the original `JsonDataSet`.
So to allow viewing the output in Kedro-Viz, I copied the curves and mapped them to another catalog entry, this time with the `JsonDataSet` type. Here, I set the filesystem arguments to read-only, to prevent writing the same output file (.json) twice.
It's a bit hacky but I can't see a reason why Kedro-Viz couldn't display the modified `PlotlyDataSet`. If we get this working we could then remove the data duplicates and the additional catalog entry.

File: pyproject.toml [hunk 6]
```diff
@@ -20,3 +20,40 @@ addopts = """
 fail_under = 0
 show_missing = true
 exclude_lines = ["pragma: no cover", "raise NotImplementedError"]
+
+[tool.poetry]
+name = "qc-data-preparation"
+version = "0.18.1"
+description = ""
+authors = ["lc3267 <melvin.strobl@kit.edu>"]
```
Comment (eileen-kuehn) [inline pyproject.toml:28]: There are some more authors involved:

```suggestion
authors = [
    {name="Eileen Kuehn", email="eileen.kuehn@kit.edu"},
    {name="Melvin Strobl", email="melvin.strobl@kit.edu"},
    {name="Christof Wendenius", email="christof.wendenius@kit.edu"},
]
```

**Review by eileen-kuehn — CHANGES_REQUESTED** [review]
There are some smaller comments and suggestions. When those are done, we are good to go 👍

File: src/qc_data_preparation/helpers/types.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+def lazy_dict(*args):
+    return {f"{arg.replace('params:','')}": arg for arg in args}
```
[the remark this answers is not in the document]
Comment (stroblme) [reply src/qc_data_preparation/helpers/types.py]: no usage; removed in [3925bdc](https://github.com/cirKITers/qc_data_preparation/pull/1/commits/3925bdc330404373c984bfb461508ca17fe16a39)
